store/tikv/lock_resolver.go

```diff
@@ -578,10 +596,258 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
+}
+
+func (lr *LockResolver) checkSecondaries(bo *Backoffer, l *Lock, curKeys [][]byte, curRegionID RegionVerID, shared *asyncResolveData) error {
+	checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+		Keys:         curKeys,
+		StartVersion: l.TxnID,
+	}
+	req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+	tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+	resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if regionErr != nil {
+		err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		logutil.BgLogger().Info("checkSecondaries region error, regrouping", zap.String("primary lock", l.String()), zap.String("region", strconv.FormatUint(curRegionID.GetID(), 10)))
+
+		// If regions have changed, then we might need to regroup the keys. Since this should be rare and for the sake
+		// of simplicity, we will resolve regions sequentially.
+		regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, curKeys, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for regionID, keys := range regions {
+			// Recursion will terminate because the resolve request succeeds or the Backoffer reaches its limit.
+			if err = lr.checkSecondaries(bo, l, keys, regionID, shared); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	if resp.Resp == nil {
+		return errors.Trace(ErrBodyMissing)
+	}
+
+	checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+	return shared.addLocks(checkResp.Locks, curRegionID, len(curKeys), l.TxnID, checkResp.CommitTs)
+}
+
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus, cleanRegions map[RegionVerID]struct{}) error {
+	tikvLockResolverCountWithResolveAsync.Inc()
+
+	regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, status.secondaries, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	shared := asyncResolveData{
+		mutex:       sync.Mutex{},
+		commitTs:    l.MinCommitTS,
+		locks:       map[RegionVerID][]*Lock{},
+		missingLock: false,
+	}
+
+	errChan := make(chan error, len(regions))
+
+	for regionID, keys := range regions {
+		curRegionID := regionID
+		curKeys := keys
+
+		go func() {
+			errChan <- lr.checkSecondaries(bo, l, curKeys, curRegionID, &shared)
+		}()
+	}
+
+	var errs []string
+	for range regions {
+		err1 := <-errChan
+		if err1 != nil {
+			errs = append(errs, err1.Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		return errors.Errorf("async commit recovery (sending CheckSecondaryLocks) finished with errors: %v", errs)
+	}
+
+	loc, err := lr.store.GetRegionCache().LocateKey(bo, l.Key)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	shared.locks[loc.Region] = append(shared.locks[loc.Region], l)
```

Comment:
It's not necessary since the lock is in `status.secondaries`.

Reply:
I assumed `l` was always the primary lock. Looking at `asyncSecondaries` in 2pc.go, we specifically exclude the primary lock from secondaries. Looking at the caller, perhaps it is called with every lock a transaction meets?

Reply:
~~Never mind. I thought `Secondaries` contains the primary key too.~~

> Looking at the caller, perhaps it is called with every lock a transaction meets?

Yes. The lock isn't needed since the `TxnStatus` contains the primary lock in https://github.com/pingcap/kvproto/pull/671.

Reply:
@nrc You can upgrade kvproto to the latest version.

Reply:
But is `l` the primary lock? I think `l` is the lock we met so that we want to recover the transaction, and the `status` was got by calling `check_txn_status` with `l.Primary` at line 473.

Reply:
Yes, the process needs to be fixed using the new proto.